### PR TITLE
Look for kademlia binary relative to coda binary

### PR DESCRIPTION
**Pull request by nholland94**

### CODE REVIEW

**Comment by enolan** [comment]
looking

File: src/lib/kademlia/membership.ml [hunk 1]
```diff
@@ -54,6 +54,43 @@ let get_project_root () =
...
+  Use /proc/PID/exe
+   - argv[0] might have been deleted (this is quite common with jenga)
+   - `cp /proc/PID/exe dst` works as expected while `cp /proc/self/exe dst` does not *)
+let get_coda_binary =
+  let our_binary_lazy =
+    lazy
+      (let open Async in
+      let open Deferred.Or_error.Let_syntax in
+      let%map os = Process.run ~prog:"uname" ~args:["-s"] () in
+      if os = "Darwin\n" then (
```
Comment (enolan) [inline src/lib/kademlia/membership.ml]: you tested this on Mac right?

File: src/lib/kademlia/membership.ml [hunk 2]
```diff
@@ -54,6 +54,43 @@ let get_project_root () =
...
+        let open Ctypes in
+        let ns_get_executable_path =
+          Foreign.foreign "_NSGetExecutablePath"
+            (ptr char @-> ptr uint32_t @-> returning void)
```
Comment (enolan) [inline src/lib/kademlia/membership.ml]: it returns an `int`. we should assert that the return value is 0.

File: src/lib/kademlia/membership.ml [hunk 3]
```diff
@@ -54,6 +54,43 @@ let get_project_root () =
...
+        in
+        let path_max = 1024 in
```
Comment (enolan) [inline src/lib/kademlia/membership.ml]: is it possible to pull in `sys/syslimits.h` with Ctypes? Hardcoding the constant makes me a bit nervous.

File: src/lib/kademlia/membership.ml [hunk 4]
```diff
@@ -54,6 +54,43 @@ let get_project_root () =
...
+        let buf = Ctypes.allocate_n char ~count:path_max in
+        let count =
+          Ctypes.allocate uint32_t (Unsigned.UInt32.of_int (path_max - 1))
```
Comment (enolan) [inline src/lib/kademlia/membership.ml]: I think this is off by one. We're passing it the size of the buffer, presumably including the null terminator

File: src/lib/kademlia/membership.ml [hunk 5]
```diff
@@ -54,6 +54,43 @@ let get_project_root () =
...
+        in
+        ns_get_executable_path buf count ;
+        let s =
+          string_from_ptr buf ~length:(!@count |> Unsigned.UInt32.to_int)
+        in
+        List.hd_exn @@ String.split s ~on:(Char.of_int 0 |> Option.value_exn) )
```
Comment (enolan) [inline src/lib/kademlia/membership.ml]: I can't believe Ctypes doesn't have helpers for dealing with C strings, but apparently it doesn't. :(

File: src/lib/kademlia/membership.ml (hunk 1 above)
Comment (bkase) [reply src/lib/kademlia/membership.ml]: @nholland94 	hook me up with instructions on how to test and I can do that for you

**Comment by nholland94** [comment]
@enolan for reference, that code for getting the binary path is existing code that works. We used that exact code snippet in our RPC parallel fork. I will still address your comments.

File: src/lib/kademlia/membership.ml (hunk 1 above)
Comment (nholland94) [reply src/lib/kademlia/membership.ml]: Yes, it is a copy-paste of existing code we use that is just scoped to another library and hidden.

File: src/lib/kademlia/membership.ml (hunk 4 above)
Comment (nholland94) [reply src/lib/kademlia/membership.ml]: I think that's true, we just hadn't hit the case where `path_len = path_max` yet.

File: src/lib/kademlia/membership.ml (hunk 5 above)
Comment (nholland94) [reply src/lib/kademlia/membership.ml]: Yes, unfortunate, I know.

File: src/lib/kademlia/membership.ml (hunk 3 above)
Comment (nholland94) [reply src/lib/kademlia/membership.ml]: Yes, it's possible. I'll just wrap a function around the preprocessor value.

**Pull request merged**